### PR TITLE
[WIP] max_pool2d without indices optimization [CPU]

### LINKED ISSUE #41099: max_pool2d always compute indices even when it's not required
https://github.com/pytorch/pytorch/blob/bdc00196d1b63a339a5b9b6aaec44cefe069a6ea/aten/src/ATen/native/Pooling.cpp#L145-L147
Proposed fix by @albanD : 
- if the user explicitly calls, it, they must have a good reason to ask for the indices.
- If they call `torch.max_pool2d`, then in the implementation of this in "ATen/native/Pooling.cpp", we should only use "max_pool2d_with_indices" if the input requires grad. (and add a new kernel that does not return the indices to use in the other case)

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43267 [WIP] max_pool2d without indices optimization [CPU]**

This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 118.4793 (1.0)
test_mkldnn_max_pool2d: 360.2836 (3.04)
test_max_pool2d_with_indices: 626.9831 (5.29)

## Discussion

The new implementation is on average 2~3 times faster than mkldnn and 5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 1
PADDING = 1
DILATION = 2
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

There is also an issue with the existing pooling implementations that they use nested at::parallel_for loops and as such only the outer most loop is parallelized since at::parallel_for does not support nesting.

Differential Revision: [D23273406](https://our.internmc.facebook.com/intern/diff/D23273406)

closes https://github.com/pytorch/pytorch/issues/28733

### CODE REVIEW

File: aten/src/ATen/native/Pooling.cpp [hunk 1]
```diff
@@ -1,16 +1,177 @@
...
+  at::parallel_for(
+      0, NB * NC * OH, 0, [&](const int64_t begin, const int64_t end) {
+        for (int64_t it = begin; it < end; ++it) {
+          const auto tensor_idx = it / OH;
+          const auto oi = it % OH;
+
+          // Compute kernel first and last rows within input bounds
+          const auto kis = ker_offset(oi * SI - PI);
+          const auto kie =
+              KH - ker_offset(((IH - 1) - (oi * SI + (KH - 1) * DI - PI)));
```
Comment (glaringlee) [inline aten/src/ATen/native/Pooling.cpp:60]: I think ((IH - 1) - (oi * SI + (KH - 1) * DI - PI)) won't < 0, right? (shape check ensured this)
ker_offset is not needed I guess. If so, we can make ker_offset inline here to save a function call.

File: aten/src/ATen/native/Pooling.cpp [hunk 2]
```diff
@@ -1,16 +1,177 @@
...
+          std::vector<scalar_t> buffer(IW + (PJ << 1), FILL);
+
+          // Copy first valid (non-padded) row to skip one loop iteration
+          const auto copy_offset = in_offset + (oi * SI + kis * DI - PI) * IW;
+          std::copy(
+              IP + copy_offset, IP + copy_offset + IW, buffer.begin() + PJ);
+
+          // Compute max column-wise for current output row
+          for (int64_t ki = kis + 1; ki < kie; ++ki) {
+            const auto ii = oi * SI + ki * DI - PI;
```
Comment (glaringlee) [inline aten/src/ATen/native/Pooling.cpp:76]: use int64_t instead auto, no need to deduce this and no need const

File: aten/src/ATen/native/Pooling.cpp [hunk 3]
```diff
@@ -1,16 +1,177 @@
...
+    const int64_t PI,
+    const int64_t PJ,
+    const int64_t DI,
+    const int64_t DJ) {
+  // Value to fill the padded region with
+  constexpr auto FILL = std::numeric_limits<scalar_t>::lowest();
+
+  // Because of padding, the kernel index may be off bounds. This computes the
+  // offset to place the kernel index in bounds.
+  const auto ker_offset = [&](const int64_t excess) {
```
Comment (glaringlee) [inline aten/src/ATen/native/Pooling.cpp:46]: no need to deduce, this is always int64_t, no need const as well

File: aten/src/ATen/native/Pooling.cpp [hunk 4]
```diff
@@ -1,16 +1,177 @@
...
+    return std::max(static_cast<int64_t>(std::ceil(excess / (DI * -1.0))), 0L);
+  };
+
+  // For each output row
+  at::parallel_for(
+      0, NB * NC * OH, 0, [&](const int64_t begin, const int64_t end) {
+        for (int64_t it = begin; it < end; ++it) {
+          const auto tensor_idx = it / OH;
```
Comment (glaringlee) [inline aten/src/ATen/native/Pooling.cpp:54]: int64_t for this line and below lines till `const auto copy_offset`. We don't need to deduce it and no need to make it const. They are always int64_t

File: aten/src/ATen/native/Pooling.cpp [hunk 5]
```diff
@@ -1,16 +1,177 @@
...
+            const auto ii = oi * SI + ki * DI - PI;
+            for (int64_t ij = 0; ij < IW; ++ij) {
+              const auto val = IP[in_offset + ii * IW + ij];
+              buffer[ij + PJ] =
+                  std::isnan(val) ? val : std::max(buffer[ij + PJ], val);
+            }
+          }
+
+          // Compute max for each cell in the current output row
+          const auto out_offset = (tensor_idx * OH + oi) * OW;
```
Comment (glaringlee) [inline aten/src/ATen/native/Pooling.cpp:85]: tensor_idx * OH  == it I think, you can use it, and add comment to clearify what 'it' is. Use int64_t instead as well.

File: aten/src/ATen/native/Pooling.cpp [hunk 6]
```diff
@@ -127,24 +288,26 @@ Tensor max_pool2d(
...
-  if(xnnpack::use_max_pool2d(self, kernel_size, padding, stride,
-                             dilation, ceil_mode)) {
+  if (xnnpack::use_max_pool2d(
+          self, kernel_size, padding, stride, dilation, ceil_mode)) {
     return xnnpack::max_pool2d(
         self, kernel_size, padding, stride, dilation, ceil_mode);
   }
 #endif
-  auto output_and_indices = at::max_pool2d_with_indices(
+  if (self.requires_grad()) {
```
Comment (glaringlee) [inline aten/src/ATen/native/Pooling.cpp:305]: I think even require grad, we can still use the new impl, anything blocking here?

**Review by glaringlee — CHANGES_REQUESTED** [review]
This is overall a great algo and easy to expand to 3d. 
Please see my comments.

**Review by glaringlee — CHANGES_REQUESTED** [review]
Some minor comments. Please rebase the code, and then I will approve it.

**Review by glaringlee — COMMENTED** [review]
@heitorschueroff 
LGTM to me now except the std::max error.
where is your benchmark?

**Review by glaringlee — CHANGES_REQUESTED** [review]

**Review by glaringlee — APPROVED** [review]
LGTM, approving. Please rebase and import to phabricator.

File: test/test_nn.py [hunk 7]
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
         helper(10, 512, 31, 31, 3, stride=2)
         helper(1, 129, 8, 8, 3, stride=2)
 
-    @onlyCUDA
```
Comment (gchanan) [inline test/test_nn.py:9813]: why do you think this was onlyCUDA before?  Isn't your test (on CPU) going to run the same thing twice and check it's the same?  That's fine I guess?

File: test/test_nn.py [hunk 8]
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
...
-        helper(1, 100000, 1, 4, ks=(1, 4))  # test for max_pool1d
+            if requires_grad:
+                y.sum().backward()
+                ref_y.sum().backward()
+                self.assertEqual(x.grad, ref_x.grad)
+
+        for requires_grad in (True, False):
+            helper(2, 8, 4, 4, 2, requires_grad)
+            helper(1, 100000, 32, 32, 4, requires_grad)
+            helper(1, 100000, 1, 4, (1, 4), requires_grad)  # test for max_pool1d
```
Comment (gchanan) [inline test/test_nn.py:9834]: is this comment wrong?

File: aten/src/ATen/native/Pooling.cpp [hunk 9]
```diff
@@ -127,24 +287,28 @@ Tensor max_pool2d(
...
-  if(xnnpack::use_max_pool2d(self, kernel_size, padding, stride,
-                             dilation, ceil_mode)) {
+  if (xnnpack::use_max_pool2d(
+          self, kernel_size, padding, stride, dilation, ceil_mode)) {
     return xnnpack::max_pool2d(
         self, kernel_size, padding, stride, dilation, ceil_mode);
   }
 #endif
-  auto output_and_indices = at::max_pool2d_with_indices(
+  if (self.requires_grad() || self.device() != at::kCPU) {
```
Comment (mruberry) [inline aten/src/ATen/native/Pooling.cpp:304]: What's up with the gradient check here? Maybe another TODO?

File: test/test_nn.py [hunk 10]
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
...
-            x = torch.randn(n, c, h, w, device='cuda', dtype=torch.float, requires_grad=True)
+        def helper(n, c, h, w, ks, requires_grad):
+            x = torch.randn(n, c, h, w, device=device, dtype=torch.float, requires_grad=requires_grad)
             ref_x = x.detach().clone().cpu().requires_grad_()
 
-            pool = torch.nn.MaxPool2d(kernel_size=ks)
-
-            y = pool(x)
-            ref_y = pool(ref_x)
+            pool = torch.nn.MaxPool2d(kernel_size=ks, return_indices=True)
```
Comment (mruberry) [inline test/test_nn.py:9818]: Would you elaborate on this change? In particular:

- In the original return_indices was not set and it default to False. 
- Doesn't your change only affect the return_indices = False codepath?

**Comment by glaringlee** [comment]
@heitorschueroff 
I put this back to [WIP] since you will add `with_indices` part as well for this PR. feel free to remove [WIP] once you are ready.

**Review by glaringlee — CHANGES_REQUESTED** [review]
@heitorschueroff
I put this back to [WIP] since you will add with_indices part as well for this PR. feel free to remove [WIP] once you are ready.

**Comment by mruberry** [comment]
Would you post your benchmark script? cc @ngimel for perf, too. Maybe a couple more sizes as a sanity check? 

- inception_v3: batch x 64 x 147 x 147
- googlenet: batch x 64 x 112 x 112

cross with params for inception v3 (kernel size 3, stride 2), googlenet (kernel size 3, stride 2, ceil mode True) and ResNet (kernel size 3, stride, padding 1)

Are there tests that the other options to maxpool2d are working correctly? Like padding, ceil mode, stride, and dilation?

File: test/test_nn.py (hunk 7 above)
Comment (xwang233) [reply test/test_nn.py:9813]: The original purpose of this onlyCUDA was to test CUDA implementation with CPU impl as a reference. I would also suggest that we keep onlyCUDA there, otherwise there would be a duplicate CPU-CPU comparison. 

In order to have a "purely CPU" test, we can have a few hard-coded pooling input and results.

File: aten/src/ATen/native/Pooling.cpp (hunk 9 above)
Comment (heitorschueroff) [reply aten/src/ATen/native/Pooling.cpp:304]: If we require grad then we need to compute indices for backward pass.

File: test/test_nn.py [hunk 11]
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
         helper(10, 512, 31, 31, 3, stride=2)
         helper(1, 129, 8, 8, 3, stride=2)
 
-    @onlyCUDA
     def test_max_pool2d(self, device):
-        def helper(n, c, h, w, ks):
-            x = torch.randn(n, c, h, w, device='cuda', dtype=torch.float, requires_grad=True)
+        def helper(n, c, h, w, ks, requires_grad):
+            x = torch.randn(n, c, h, w, device=device, dtype=torch.float, requires_grad=requires_grad)
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply test/test_nn.py:9815]: x.to('cpu', copy=True).requires_grad_() is returning None for some reason.

[the remark this answers is not in the document]
Comment (heitorschueroff) [reply test/test_nn.py:9815]: or rather calling .grad on the returning Tensor is None

**Review by xwang233 — COMMENTED** [review]
Besides your changes to `test_max_pool2d`, there is also another `test_max_pool2d_indices`. Would you mind combine the two tests together, rather than modifying the current `test_max_pool2d` to another duplicates of "test_another_max_pool2d_indices"? Thanks!

https://github.com/pytorch/pytorch/blob/42a53607359af5c84638318d30b939f48c6aab8e/test/test_nn.py#L9868-L9889
